README

============================================================================

    USE THIS SOFTWARE AT YOUR OWN RISK.  I TAKE NO RESPONSIBILITY
    FOR ANY DAMAGE THAT MAY RESULT FROM USING THIS SOFTWARE.

============================================================================

Samsung Uproar Linux Utility (sulu)

Sulu is a graphical utility for transferring mp3's between
your Samsung Uproar and your Linux computer.  It uses the 
GTK+ toolkit for the user interface and libusb.  Sulu is
licensed under the GPL (see the file COPYING).

You will need a kernel with USB support properly configured.  
I used a stock RedHat 7.2 install.  You can check whether USB 
is properly configured for your hardware by entering:

grep Vendor /proc/bus/usb/devices

You should see a line with "Vendor=04e8 ProdID=6600".

Once USB is properly setup you will have to install the 
shared library and the devel libraries for libusb.
(http://libusb.sourceforge.net/download.html).

To compile sulu, just type "make".  The executable is named sulu.
You need to run as root.

 
Author: 
kal@users.sourceforge.net

Credits: 
Big thanks goes to Kasima Tharnpipitchai for his work on the 
protocol and Johannes Erdfelt for libusb.  



changelog
0.04 added a notes directory, added a todo file, clarified the README
0.03 have to reload the playlist after a download to generate new ids
0.02 convert non-isprint chars to space in playlist 
     timeout callbacks in UI to show a status line when doing USB calls
0.01 initial alpha quality release


main.h

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/
 

#ifndef main_h
#define main_h

#define MAX_SYSTEM_FILES 1024
#define MAX_FILENAME_LEN 1024
#define STARTUP_X_SIZE   520
#define STARTUP_Y_SIZE   360
#define DEVICE_PANE_X    (STARTUP_X_SIZE / 2)
#define DEVICE_PANE_Y    (STARTUP_Y_SIZE)
#define SYSTEM_PANE_X    (STARTUP_X_SIZE - DEVICE_PANE_X)
#define SYSTEM_PANE_Y    (STARTUP_Y_SIZE)
#define GUTTER_SIZE      20
#define MIN_PANE_WIDTH   200

gint Delete_event(GtkWidget * widget, GdkEvent * event, gpointer data);

GdkColor *FgDir;
GdkColor *FgFile;

#endif


lib_uproar.h

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/

 
#ifndef lib_uproar_h
#define lib_uproar_h

#include <usb.h>

#define UPROAR_VENDOR   0x04E8
#define UPROAR_PRODUCT  0x6600

#define UPROAR_SUCCESS                0
#define UPROAR_ERROR                 -1
#define UPROAR_DEVICE_NOT_FOUND      -1000
#define UPROAR_NOT_ENOUGH_SPACE      -1010

extern struct usb_device *Uproar_dev;

int Uproar_initialize();
unsigned char *Uproar_get_device_info();
int Uproar_load_playlist();
unsigned char *Uproar_get_track_info(int tracknum, int *filesize);
int Uproar_get_capacity();
int Uproar_get_available();
int Uproar_num_tracks();
int Uproar_delete_track(int tracknum);
int Uproar_reformat_memory();
int Uproar_download(unsigned char *filepath);


#endif


device_pane.c

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/
 

#include <stdio.h>
#include <dirent.h>
#include <malloc.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <gtk/gtk.h>

#include <usb.h>

#include "main.h"
#include "device_pane.h"
#include "lib_uproar.h"

GtkWidget *Create_device_pane()
{
    GtkWidget *vertical_box;
    GtkWidget *scrolled_window;
    gchar *titles[2] = { "Filename", "MB" };
    int padding = 2;

    GtkWidget *menu_bar;

    GtkWidget *file_menu;
    GtkWidget *file_item;
    GtkWidget *selectall_item;
    GtkWidget *unselectall_item;
    GtkWidget *invert_item;

    GtkWidget *action_menu;
    GtkWidget *action_item;
    GtkWidget *refresh_item;
    GtkWidget *delete_item;
    GtkWidget *reformat_item;

    /* 
     * create a vertical box to hold everything in this pane 
     */
    vertical_box = gtk_vbox_new(FALSE, padding);
    gtk_widget_show(vertical_box);

    /* 
     * create the selection menu and link it with the top level item 
     */
    file_menu = gtk_menu_new();
    file_item = gtk_menu_item_new_with_label("File");
    gtk_widget_show(file_item);
    gtk_menu_item_set_submenu(GTK_MENU_ITEM(file_item), file_menu);

    /* selection->selectall */
    selectall_item = gtk_menu_item_new_with_label("Select all");
    gtk_menu_append(GTK_MENU(file_menu), selectall_item);
    gtk_signal_connect_object(GTK_OBJECT(selectall_item), "activate",
			      GTK_SIGNAL_FUNC(Selectall_device_pane),
			      (gpointer) "Selectall_device_pane");
    gtk_widget_show(selectall_item);

    /* selection->unselectall */
    unselectall_item = gtk_menu_item_new_with_label("Unselect all");
    gtk_menu_append(GTK_MENU(file_menu), unselectall_item);
    gtk_signal_connect_object(GTK_OBJECT(unselectall_item), "activate",
			      GTK_SIGNAL_FUNC(Unselectall_device_pane),
			      (gpointer) "Unselectall_device_pane");
    gtk_widget_show(unselectall_item);

    /* selection->invert */
    invert_item = gtk_menu_item_new_with_label("Invert Selection");
    gtk_menu_append(GTK_MENU(file_menu), invert_item);
    gtk_signal_connect_object(GTK_OBJECT(invert_item), "activate",
			      GTK_SIGNAL_FUNC(Invert_device_pane),
			      (gpointer) "Invert_device_pane");
    gtk_widget_show(invert_item);

    /* 
     * create the transfer menu and link it with the top level item 
     */
    action_menu = gtk_menu_new();
    action_item = gtk_menu_item_new_with_label("Action");
    gtk_widget_show(action_item);
    gtk_menu_item_set_submenu(GTK_MENU_ITEM(action_item), action_menu);

    /* action->refresh */
    refresh_item = gtk_menu_item_new_with_label("Refresh List");
    gtk_menu_append(GTK_MENU(action_menu), refresh_item);
    gtk_signal_connect_object(GTK_OBJECT(refresh_item), "activate",
			      GTK_SIGNAL_FUNC(Refresh_device_pane),
			      (gpointer) "Refresh_device_pane");
    gtk_widget_show(refresh_item);

    /* action->delete tracks */
    delete_item = gtk_menu_item_new_with_label("Delete tracks");
    gtk_menu_append(GTK_MENU(action_menu), delete_item);
    gtk_signal_connect_object(GTK_OBJECT(delete_item), "activate",
			      GTK_SIGNAL_FUNC(Delete_tracks),
			      (gpointer) "Delete_tracks");
    gtk_widget_show(delete_item);

    /* action->reformat memory */
    reformat_item = gtk_menu_item_new_with_label("Reformat memory");
    gtk_menu_append(GTK_MENU(action_menu), reformat_item);
    gtk_signal_connect_object(GTK_OBJECT(reformat_item), "activate",
			      GTK_SIGNAL_FUNC(Reformat_memory),
			      (gpointer) "Reformat_memory");
    gtk_widget_show(reformat_item);

    /* 
     * create the menu bar and add the top level items 
     */
    menu_bar = gtk_menu_bar_new();
    gtk_menu_bar_append(GTK_MENU_BAR(menu_bar), file_item);
    gtk_menu_bar_append(GTK_MENU_BAR(menu_bar), action_item);

    /* set the packing policy for resizes */
    gtk_container_add(GTK_CONTAINER(vertical_box), menu_bar);
    gtk_box_set_child_packing(GTK_BOX(vertical_box), menu_bar, FALSE,
			      FALSE, padding, GTK_PACK_START);
    gtk_widget_show(menu_bar);

    /*
     * create a status line
     */
    Status_device = gtk_label_new("");

    /* set the packing policy for resizes */
    gtk_container_add(GTK_CONTAINER(vertical_box), Status_device);
    gtk_box_set_child_packing(GTK_BOX(vertical_box), Status_device, FALSE,
			      FALSE, padding, GTK_PACK_START);
    gtk_widget_show(Status_device);

    /* 
     * create a new scrolled window, with scrollbars only if needed
     */
    scrolled_window = gtk_scrolled_window_new(NULL, NULL);
    gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scrolled_window),
				   GTK_POLICY_AUTOMATIC,
				   GTK_POLICY_AUTOMATIC);
    gtk_widget_set_usize(GTK_WIDGET(scrolled_window), SYSTEM_PANE_X,
			 SYSTEM_PANE_Y);
    gtk_container_add(GTK_CONTAINER(vertical_box), scrolled_window);
    gtk_widget_show(scrolled_window);

    /* create a new clist and put it in the scrolled window */
    Clist_device = gtk_clist_new_with_titles(2, titles);
    gtk_clist_set_selection_mode(GTK_CLIST(Clist_device),
				 GTK_SELECTION_MULTIPLE);
    gtk_clist_set_column_width(GTK_CLIST(Clist_device), 0,
			       MIN_PANE_WIDTH + 10);
    gtk_clist_set_column_width(GTK_CLIST(Clist_device), 1, 10);
    gtk_clist_column_titles_passive(GTK_CLIST(Clist_device));

    /* add the file list to the window */
    Refresh_device_pane();

    gtk_container_add(GTK_CONTAINER(scrolled_window), Clist_device);
    gtk_widget_show(Clist_device);

    return vertical_box;
}

void Selectall_device_pane()
{

    gtk_clist_select_all(GTK_CLIST(Clist_device));
}

void Unselectall_device_pane()
{

    gtk_clist_unselect_all(GTK_CLIST(Clist_device));
}

void Invert_device_pane()
{
    int selected;
    gint i;
    gint row = -1;
    gint total_rows = GTK_CLIST(Clist_device)->rows;
    GList *glist;

    for (i = 0; i < total_rows; i++) {
	selected = 0;
	glist = GTK_CLIST(Clist_device)->selection;
	while (glist != NULL) {
	    row = (gint) glist->data;
	    if (row == i) {
		selected = 1;
		break;
	    }
	    glist = glist->next;
	}
	if (selected == 0) {
	    gtk_clist_select_row(GTK_CLIST(Clist_device), i, 0);
	} else {
	    gtk_clist_unselect_row(GTK_CLIST(Clist_device), i, 0);
	}
    }
}

void Delete_tracks()
{
    gtk_label_set_text(GTK_LABEL(Status_device), "DELETING...");
    gtk_timeout_add(10, Timeout_device_delete, 0);
}

gint Timeout_device_delete(gpointer gdata)
{
    int retval;
    int selected;
    gint i;
    gint row = -1;
    gint total_rows;
    GList *glist;

    total_rows = GTK_CLIST(Clist_device)->rows;
    for (i = 0; i < total_rows; i++) {
	selected = 0;
	glist = GTK_CLIST(Clist_device)->selection;
	while (glist != NULL) {
	    row = (gint) glist->data;
	    if (row == i) {
		selected = 1;
		break;
	    }
	    glist = glist->next;
	}
	if (selected) {
	    retval = Uproar_delete_track(i);
	    if (retval != UPROAR_SUCCESS) {
                fprintf(stderr, "\nError deleting file from device.");
	    }
	}

    }
    gtk_clist_unselect_all(GTK_CLIST(Clist_device));
    Refresh_device_pane();
    return FALSE;
}


void Reformat_memory()
{
    gtk_label_set_text(GTK_LABEL(Status_device), "REFORMATING...");
    gtk_timeout_add(10, Timeout_device_reformat, 0);
}

gint Timeout_device_reformat(gpointer gdata)
{
    int retval;

    retval = Uproar_reformat_memory();
    if (retval != UPROAR_SUCCESS) {
        fprintf(stderr, "\nError reformatting device.");
    }
    gtk_clist_unselect_all(GTK_CLIST(Clist_device));
    Refresh_device_pane();
    return FALSE;
}

void Refresh_device_pane()
{
    int retval;
    int tracknum;
    char *trackname;
    gchar *clist_txt[2];
    int filesize;
    float megabytes;
    char status[80];

    /* load the playlist */
    retval = Uproar_load_playlist();
    if (retval != UPROAR_SUCCESS) {
	gtk_label_set_text(GTK_LABEL(Status_device),
			   "Error loading playlist.");
	return;
    }

    gtk_clist_freeze(GTK_CLIST(Clist_device));

    gtk_clist_clear(GTK_CLIST(Clist_device));

    tracknum = 0;
    while (1) {
	trackname = Uproar_get_track_info(tracknum, &filesize);
	if (trackname == 0) {
	    break;
	}
	megabytes = ((float) filesize) / ((float) 1024 * 1024);
	clist_txt[0] = g_strdup_printf("%s", trackname);
	clist_txt[1] = g_strdup_printf("%.2f", megabytes);
	gtk_clist_append(GTK_CLIST(Clist_device), clist_txt);
	g_free(clist_txt[0]);
	g_free(clist_txt[1]);
	tracknum++;
    }

    gtk_clist_thaw(GTK_CLIST(Clist_device));

    megabytes = ((float) Uproar_get_available()) / ((float) 1024 * 1024);
    sprintf(status, "Space Available %.2fMB", megabytes);
    gtk_label_set_text(GTK_LABEL(Status_device), status);
}



device_pane.h

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/


#ifndef device_pane_h
#define device_pane_h

#include <gtk/gtk.h>

GtkWidget *Create_device_pane();
GtkWidget *Clist_device;
GtkWidget *Status_device;
void Refresh_device_pane();
void Selectall_device_pane();
void Unselectall_device_pane();
void Invert_device_pane();
void Delete_tracks();
void Reformat_memory();

gint Timeout_device_delete(gpointer data);
gint Timeout_device_reformat(gpointer data);

#endif


system_pane.h

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/
 

#ifndef system_pane_h
#define system_pane_h

#include <gtk/gtk.h>

GtkWidget *Create_system_pane();
GtkWidget *Clist_system;
GtkWidget *Status_system;
void Refresh_system_pane();
void Selectall_system_pane();
void Unselectall_system_pane();
void Invert_system_pane();
void Transfer_system_pane();
gint Is_regular_file(gint i);
void Select_row_callback(GtkWidget * widget, gint row, gint col,
			 GdkEventButton * ev, gpointer data);

gint Timeout_system_download(gpointer data);

#endif


main.c

/*

    This file is part of the Samsung Uproar Linux Utility (sulu).
 
    Sulu is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    Sulu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with sulu.  See the file COPYING.  If you haven't received
    a copy of the GNU General Public License, write to:
 
        Free Software Foundation, Inc., 
        59 Temple Place, Suite 330, Boston, MA  
        02111-1307  USA
 
  kal@users.sourceforge.net
 
*/
 

#include <stdio.h>
#include <gtk/gtk.h>
#include <usb.h>

#include "main.h"
#include "system_pane.h"
#include "device_pane.h"
#include "lib_uproar.h"

gint Delete_event(GtkWidget * widget, GdkEvent * event, gpointer data)
{
    gtk_main_quit();
    return (FALSE);
}

int main(int argc, char **argv)
{
    GtkWidget *window;
    GtkWidget *vertical_layout;
    GtkWidget *hpaned;
    GtkWidget *frame1;
    GtkWidget *frame2;
    GtkWidget *system_pane;
    GtkWidget *device_pane;
    GdkColormap *cmap;
    GdkColor col1 = { 0, 0x1000, 0x7000, 0x1000 };
    GdkColor col2 = { 0, 0x7000, 0x1000, 0x1000 };
    int retval;
    char *devstr;

    retval = Uproar_initialize();
    if (retval < 0) {
	fprintf(stderr, "\n\nNO MP3 DEVICE FOUND.\n\n");
	exit(-1);
    }
    devstr = Uproar_get_device_info();
    if (devstr == 0) {
	fprintf(stderr, "\n\nNO MP3 DEVICE STRING.\n\n");
	exit(-1);
    }

    /* initialize the gtk */
    gtk_init(&argc, &argv);

    /* setup the default style */
    gtk_rc_parse(".sulurc");

    /* get some colors */
    cmap = gdk_colormap_get_system();
    if (gdk_colormap_alloc_color(cmap, &col1, FALSE, TRUE) &&
	gdk_colormap_alloc_color(cmap, &col2, FALSE, TRUE)) {
	FgFile = &col1;
	FgDir = &col2;
    } else {
	fprintf(stderr, "\nNO COLORS");
	FgDir = 0;
	FgFile = 0;
    }

    /* create the main window */
    window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_widget_set_name(window, "main_window");
    gtk_window_set_title(GTK_WINDOW(window),
			 "Samsung Uproar Linux Utility (sulu) Version 0.04");
    gtk_signal_connect(GTK_OBJECT(window), "destroy",
		       GTK_SIGNAL_FUNC(Delete_event), NULL);
    gtk_container_set_border_width(GTK_CONTAINER(window), 0);
    gtk_widget_set_usize(GTK_WIDGET(window), STARTUP_X_SIZE,
			 STARTUP_Y_SIZE);

    /* create a vertical box layout */
    vertical_layout = gtk_vbox_new(FALSE, 5);
    gtk_container_add(GTK_CONTAINER(window), vertical_layout);
    gtk_widget_show(vertical_layout);

    /* create the side by side pane */
    hpaned = gtk_hpaned_new();
    gtk_paned_set_gutter_size(GTK_PANED(hpaned), GUTTER_SIZE);
    gtk_widget_set_usize(hpaned, 2 * MIN_PANE_WIDTH + GUTTER_SIZE, -1);
    gtk_container_add(GTK_CONTAINER(vertical_layout), hpaned);
    gtk_widget_show(hpaned);

    /* create the system pane, left side */
    system_pane = Create_system_pane();
    gtk_widget_set_name(system_pane, "system_pane");
    frame1 = gtk_frame_new(NULL);
    gtk_paned_pack1(GTK_PANED(hpaned), frame1, TRUE, FALSE);
    gtk_widget_set_usize(frame1, MIN_PANE_WIDTH, -1);
    gtk_widget_show(frame1);
    gtk_container_add(GTK_CONTAINER(frame1), system_pane);
    gtk_widget_show(system_pane);

    /* create the device pane, right side */
    device_pane = Create_device_pane();
    gtk_widget_set_name(device_pane, "device_pane");
    frame2 = gtk_frame_new(NULL);
    gtk_paned_pack2(GTK_PANED(hpaned), frame2, TRUE, FALSE);
    gtk_widget_set_usize(frame2, MIN_PANE_WIDTH, -1);
    gtk_widget_show(frame2);
    gtk_container_add(GTK_CONTAINER(frame2), device_pane);
    gtk_widget_show(device_pane);

    // gtk_paned_add2(GTK_PANED(hpaned), device_pane);
    // gtk_widget_show(device_pane);

    /* show the window and start gtk processing */
    gtk_widget_show(window);
    gtk_main();

    return 0;
}
